File: 算法/Java/第3章 排序算法/3.7 计数排序/CountingSort.java
import java.util.Arrays;

public class CountingSort {
    public static void main(String[] args) {
        int[] arr = {3, 4, 3, 2, 1, 2, 6, 5, 4, 7};
        countingSort(arr);
        System.out.println(Arrays.toString(arr));
    }

    public static void countingSort(int[] arr) {
        int max = arr[0];
        int min = arr[0];
        for(int i = 1; i < arr.length; i++) {
            if(arr[i] > max) {
                max = arr[i];
            }
            if(arr[i] < min) {
                min = arr[i];
            }
        }

        int range = max - min + 1;
        int[] table = new int[range];

        for(int i = 0; i < arr.length; i++) {
            table[arr[i] - min]++;
        }

        int cnt = 0;
        for(int i = 0; i < range; i++) {
            while(table[i]-- > 0) {
                arr[cnt++] = i + min;
            }
        }
    }
}

File: code/Python/第8章 排序算法/8.7 计数排序/counting_sort.py
def counting_sort(lst):
    n = len(lst)
    max_elem = max(lst)
    min_elem = min(lst)
    length = max_elem - min_elem + 1
    table = [0] * length

    for i in range(n):
        table[lst[i] - min_elem] += 1
    
    cnt = 0
    for i in range(length):
        while table[i]:
            lst[cnt] = i + min_elem
            cnt += 1
            table[i] -= 1

def main():
    lst = [3, 4, 3, 2, 1, 2, 6, 5, 4, 7]
    counting_sort(lst)
    print(lst)

if __name__ == "__main__":
    main()